src/dali/plugin/pair_converter/ccxt.py

```diff
@@ -0,0 +1,326 @@
+
+# Native format keywords
+_ID: str = "id"
+_BASE: str = "base"
+_QUOTE: str = "quote"
+
+# Time in ms
+_MINUTE: str = "1m"
+_FIVE_MINUTE: str = "5m"
+_FIFTEEN_MINUTE: str = "15m"
+_ONE_HOUR: str = "1h"
+_FOUR_HOUR: str = "4h"
+_ONE_DAY: str = "1d"
+_TIME_GRANULARITY: List[str] = [_MINUTE, _FIVE_MINUTE, _FIFTEEN_MINUTE, _ONE_HOUR, _FOUR_HOUR, _ONE_DAY]
+_TIME_GRANULARITY_IN_SECONDS: List[int] = [60, 300, 900, 3600, 14400, 86400]
+
+# Currently supported exchanges
+_BINANCE: str = "Binance.com"
+_KRAKEN: str = "Kraken"
+_LIQUID: str = "Liquid"
+_FIAT_EXCHANGE: str = "Exchangerate.host"
+_EXCHANGE_DICT: Dict[str, Any] = {_BINANCE: binance, _KRAKEN: kraken, _LIQUID: liquid}
+
+# Delay in fractional seconds before making a request to avoid too many request errors
+# Kraken states it has a limit of 1 call per second, but this doesn't seem to be correct.
+# It appears Kraken public API is limited to around 12 calls per minute.
+# There also appears to be a limit of how many calls per 2 hour time period.
+# Being authenticated lowers this limit.
+_REQUEST_DELAYDICT: Dict[str, float] = {_BINANCE: 0.0, _KRAKEN:5.1, _LIQUID:0}
+
+# Alternative Markets and exchanges for stablecoins or untradeable assets
+_ALTMARKET_EXCHANGES_DICT: Dict[str, str] = {"USDTUSD": _KRAKEN, "SOLOXRP": _LIQUID}
+_ALTMARKET_BY_BASE_DICT: Dict[str, str] = {"USDT": "USD", "SOLO": "XRP"}
+
+# Time constants
+_MS_IN_SECOND: int = 1000
+
+
+class StableCoin(NamedTuple):
+    symbol: str
+    fiat_code: str
+
+
+class AssetPairAndTimestamp(NamedTuple):
+    timestamp: datetime
+    from_asset: str
+    to_asset: str
+    exchange: str
+
+
+class AssetPairAndHistoricalPrice(NamedTuple):
+    from_asset: str
+    to_asset: str
+    exchange: str
+    historical_data: Optional[HistoricalBar] = None
+
+
+class PairConverterPlugin(AbstractPairConverterPlugin):
+    def __init__(self, historical_price_type: str) -> None:
+        super().__init__(historical_price_type=historical_price_type)
+        self.__logger: logging.Logger = create_logger(f"{self.name()}/{historical_price_type}")
+
+        self.__exchanges: Dict[str, Exchange] = {}
+        self.__exchange_markets: Dict[str, Dict[str, List[str]]] = {}
+        self.__exchange_graphs: Dict[str, Dict[str, List[str]]] = {}
+        self.__exchange_last_request: Dict[str, float] = {}
+
+    def name(self) -> str:
+        return "CCXT-converter"
+
+    def cache_key(self) -> str:
+        return self.name()
+
+    @property
+    def exchanges(self) -> Dict[str, Exchange]:
+        return self.__exchanges
+
+    @property
+    def exchange_markets(self) -> Dict[str, Dict[str, List[str]]]:
+        return self.__exchange_markets
+
+    @property
+    def exchange_graphs(self) -> Dict[str, Dict[str, List[str]]]:
+        return self.__exchange_graphs
+
+    def _bfs_cyclic(self, graph: Dict[str, List[str]], start: str, end: str) -> Optional[List[str]]:
+
+        # maintain a queue of paths
+        queue: List[List[str]] = []

```

Comment:
You may be able to avoid this nested list and just use one Vertex queue (similar to `visited`) and one Dict capturing the parent of any given node in the spanning tree. But no need to do this now: we can leave this as a future improvement.